File: NativeBridge/NativeAPIAccessDemo/android/app/src/main/java/com/nativeapiaccessdemo/NativeAPIModule.java
package com.nativeapiaccessdemo;

import android.support.annotation.Nullable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;

import java.util.Map;
import java.util.HashMap;

@ReactModule(name = "NativeAPIModule")
public class NativeAPIModule extends ReactContextBaseJavaModule {
        
    public NativeAPIModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "NativeAPIModule";
    }

    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("FIRST_DAY_OF_WEEK", "Monday");
        constants.put("FIRST_MONTH_OF_YEAR", "January");
        return constants;
    }

    @ReactMethod
    public void addNumbers(final double a, final double b, Callback callback) {
        double result = a + b;
        callback.invoke(result);
    }

    @ReactMethod
    public void makeGreetingText(final String name, Promise promise) {
        String errorCode = "1";
        String errorMessage = "name can't be null or empty in makeGreetingText.";

        if( name == null || name.isEmpty() || (name.trim().isEmpty()) ) {
            promise.reject(errorCode, errorMessage);
        } else {
            promise.resolve(String.format("Hello %s, Welcome to Native Module Demo.", name));            
        }
    }

    // Event Emitter Helper
    private void sendEvent(String eventName, @Nullable WritableMap params) {
        getReactApplicationContext()
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
            .emit(eventName, params);
    }

    // Manually generate event
    @ReactMethod
    public void generateEvent() {
        WritableMap params = Arguments.createMap();
        params.putString("message", "I am from generated event.");
        
        // Call helper to send event
        sendEvent("onGenerateEvent", params);
    }
}

File: NativeBridge/NativeAPIAccessDemo/ios/NativeBridges/NativeAPIModule.m
//
//  NativeAPIModule.m
//  NativeAPIAccessDemo
//
//  Created by Mohammad Kamrul Hasan on 24/2/18.
//  Copyright © 2018 Facebook. All rights reserved.
//

#import "NativeAPIModule.h"

@implementation NativeAPIModule

RCT_EXPORT_MODULE();

// Constants
- (NSDictionary *)constantsToExport
{
  return @{
           @"FIRST_DAY_OF_WEEK": @"Monday",
           @"FIRST_MONTH_OF_YEAR": @"January",
          };
}

RCT_EXPORT_METHOD(addNumbers:(double)a  with:(double)b callback:(RCTResponseSenderBlock)callback)
{
  double sum = a + b;
  callback(@[@(sum)]);
}

RCT_EXPORT_METHOD(makeGreetingText:(NSString *)name resolver: (RCTPromiseResolveBlock)resolve rejecter: (RCTPromiseRejectBlock)reject)
{
  NSString *errorCode = @"1";
  NSString *errorMessage = @"name can't be null or empty in makeGreetingText.";
  
  if( (name == (id) [NSNull null] || [name length] == 0 || [name isEqual:@""])) {
    reject(errorCode, errorMessage, nil);
  } else {
    NSString *trimmedName = [name stringByTrimmingCharactersInSet: [NSCharacterSet whitespaceCharacterSet]];
    if([trimmedName length] == 0 || [trimmedName isEqual:@""]) {
      reject(errorCode, errorMessage, nil);
    } else {
      NSString *greetingText = [NSString stringWithFormat:@"Hello %@, Welcome to Native Module Demo.", trimmedName];
      resolve(greetingText);
    }
  }
}

// Event Emitters
- (NSArray<NSString *> *)supportedEvents
{
  return @[@"onGenerateEvent"];
}

// Manually trigger an event
RCT_EXPORT_METHOD(generateEvent)
{
  [self sendEventWithName:@"onGenerateEvent" body:@{@"message": @"I am from generated event."}];
}

@end
